### PR TITLE
dataset: fix dataset performance drops signifcantly on high dimensions data.

### PR DESCRIPTION
**Pull request by pissang**
## Brief Information

This pull request is in the type of:

- [ ] bug fixing
- [ ] new feature
- [x] others

### What does this PR do?

This PR refactors the `List` module and aims to fix the terrible performance when many series are sharing the same high dimension dataset.

#### Result of test case on my MacBook

1. Case from #11907. 1000 dimensions dataset :
Before: About 45s to render.
After: About 300ms to render and only takes 20MB heap memory.

2. test/sample-compare.html. 1 million, 4 dimensions dataset with LTTB sampling
Before: About 40ms
After: About 25ms

### Fixed issues

#11907

### Details explain

The major reason why high dimension data is so slow previously is that echarts will process all dimensions and recreate storage from the dataset in each series. In most cases, it won't cost much. But in #11907, There is extra high dimension data and each dimension will be distributed to a line series. In this case, echarts need to process the 1000 dimensions 1000 times. It brings a lot of pressure to the GC.

So the main purpose of this refactoring is quite clear: This dataset which is shared by all the series should only be processed once. To achieve this purpose, we need to divide the original `List` module into two modules: 

+ DataStorage: process and store the multi-dimension data. It can be shared by multiple series.
+ SeriesData: A data reading wrapper of the storage for each series. Also manages the relationship of data and graphic info(visual, elements). The interface is almost the same as the original `List` to avoid breaks.

It sounds quite straightforward and simple. So why we didn't do it before? Because we need to collect enough information before parsing data and it's hard to prepare the parsed data for sharing ahead. For example, we need to know which axis of a cartesian coordinate system is a category, so we can parse data of this dimension as category data instead of converting it to numbers. We also detect the type from data. But we don't trust it. Because detect every value is costly, we can only detect part of them. Also, we can accept a number as a category, a string as a date, a string number('123') as a number, etc. So the easiest way is parsing the source when initializing each series after all necessary information is prepared.

In this PR keep the idea of preparing the storage ahead based on the information dataset given. If series needs the source to be parsed differently. We discard the shared storage and create a new one for this series. There are several factors that will invalidate the shared storage:

1. series.seriesLayoutBy is different. 
2. series.sourceHeader is different.
3. series also specify dimensions and it's different from the dataset.
4. dataset detects one dimension to be number but the series needs it to be ordinal.
5. dataset detects one dimension to be ordinal but the series needs it to be number or time.

This invalidate strategy will ensure the behavior is correct but may have extra costs in some cases. So there is a best practice to avoid this:

**Always specify the name and type of dimensions in the dataset.**

### CODE REVIEW

**Review by 100pah — DISMISSED** [review]

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/model/Series.ts [hunk 1]
```diff
@@ -385,14 +391,22 @@ class SeriesModel<Opt extends SeriesOption = SeriesOption> extends ComponentMode
         inner(this).data = data;
     }
 
-    getSource(): Source {
-        return inner(this).sourceManager.getSource();
+    getEncode() {
+        const encode = (this as Model<SeriesEncodeOptionMixin>).get('encode');
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/model/Series.ts:395]: Fixed

File: src/data/helper/sourceManager.ts [hunk 2]
```diff
@@ -346,8 +350,79 @@ export class SourceManager {
...
+        if (!cachedStore) {
+            const upSourceMgr = this._getUpstreamSourceManagers()[0];
+
+            if (isSeries(this._sourceHost) && upSourceMgr) {
+                cachedStore = upSourceMgr._innerGetDataStorage(endSource, sourceReadKey);
+            }
+            else {
+                const dimensionsDefine = source.dimensionsDefine;
+                // Can't create a store if don't know dimension..
+                if (source && dimensionsDefine) {
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/sourceManager.ts]: Yes, the code here may mislead the readers and it should be endSource. I've changed it a bit. 

Basically, we want the storage to be always created from the source of series, which has the most correct read settings and managed/shared in the upstream source manager.

File: src/data/helper/sourceManager.ts [hunk 3]
```diff
@@ -346,8 +350,79 @@ export class SourceManager {
...
      *                    Most cases there is only one source.
      */
-    getSource(sourceIndex?: number) {
-        return this._sourceList[sourceIndex || 0];
+    getSource(sourceIndex?: number): Source {
+        sourceIndex = sourceIndex || 0;
+        const source = this._sourceList[sourceIndex];
+        if (!source) {
+            // Series may share source instance with dataset.
+            const upSourceMgrList = this._getUpstreamSourceManagers();
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/sourceManager.ts:358]: This optimization is based on the profile result. In fact, in the case of #11907, The most memory cost thing is the storage of dimensions info. Previously we need to store 1000 dimensions in each series, which is 1 million in total. It's not only in the `List` instance. But also in the `Source` instance.

`createSource` will also parse all dimensions and store them in the `Source` instance. If we don't share them, the heap memory will increase from 20MB to 140MB and the render time will also be much longer. What makes it worse is it increases exponentially.

Sharing source may make things bit more complex. But I think it won't break the design of the dataset. And the good thing is it can be all done in the SourceManager and won't leak the abstraction to other modules.

File: src/data/helper/createDimensions.ts [hunk 4]
```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
...
+
+    const sysDims = opt.coordDimensions || [];
+    const dimsDef = opt.dimensionsDefine || source.dimensionsDefine || [];
+    const coordDimNameMap = createHashMap<true, DimensionName>();
+    const result: DataDimensionInfo[] = [];
+    const omitUnusedDimensions = opt.omitUnusedDimensions;
+    const isUsingSourceDimensionsDef = dimsDef === source.dimensionsDefine;
+    // Try to cache the dimNameMap if the dimensionsDefine is from source.
+    const canCacheDimNameMap = (isUsingSourceDimensionsDef && omitUnusedDimensions);
+    let dataDimNameMap = canCacheDimNameMap && inner(source).dimNameMap;
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:112]: Added.

File: src/data/helper/createDimensions.ts [hunk 5]
```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
...
- * @param opt.encodeDefine By default `source.encodeDefine` Overwrite source define.
- * @param opt.encodeDefaulter Make default encode if user not specified.
+ * This method builds the relationship between:
+ * + "what the coord sys or series requires (see `sysDims`)",
+ * + "what the user defines (in `encode` and `dimensions`, see `opt.dimsDef` and `opt.encodeDef`)"
+ * + "what the data source provids (see `source`)".
+ *
+ * Some guess strategy will be adapted if user does not define something.
+ * If no 'value' dimension specified, the first no-named dimension will be
+ * named as 'value'.
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:85]: Fixed

File: src/chart/helper/createSeriesDataFromArray.ts [hunk 6]
```diff
@@ -107,26 +99,92 @@ function createListFromArray(source: Source | OptionSourceData, seriesModel: Ser
...
+            dimInfoList = createDimensions(sourceOrStore, zrUtil.extend(createDimensionOptions, {
+                omitUnusedDimensions: true
+            }));
+            // Fallback
+            firstCategoryDimIndex = injectOrdinalMeta(
+                dimInfoList, opt.createInvertedIndices, coordSysInfo
+            );
+            sourceOrStore = source;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/chart/helper/createSeriesDataFromArray.ts:164]: This code will try to create a dimensions list that only include needed and use `syncDimensionTypes(dimInfoList)` to sync to the storage. If fails to sync, it will recreate a full dimensions list which will be used when recreating the DataStorage. I updated the comment a bit to explain it.

File: src/data/helper/sourceManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (100pah) [reply src/data/helper/sourceManager.ts:358]: Oh, get it.
Do we need to add some brief comment or the link of this explantation in source code? Other one may not get the reason of this consideration when reading code.

File: src/data/SeriesData.ts [hunk 7]
```diff
@@ -0,0 +1,1338 @@
...
+            : cb) as any
+        );
+        return list;
+    }
+
+    /**
+     * Large data down sampling on given dimension
+     * @param sampleIndex Sample index for name and id
+     */
+    downSample(
```
Comment (100pah) [inline src/data/SeriesData.ts:912]: ### Just a programming pattern thinking

Since we have shared data in `DataStorage` instead of `SeriesData/List`, 
should we change the interface like `SeriesData#downSamle`/`SeriesData#lttbDownSample`/`SeriesData#map` from immutable to mutable (only provide immutable interface in `DataStorage`),
And substitute `SeriesData#reset` for `SeriesData#cloneShallow`.
And remove `Series#setData`.
Finally a `series` instance will not change the reference of `seriesData` instance,
and the file `src/data/helper/linkSeriesData.ts`, which is a little weird, can be removed.
And the concept might be more simple: only `DataStorage` can be shared and immutable, rather than both `DataStorage` and `SeriesData` need to be  immutable.

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/data/helper/createDimensions.ts [hunk 8]
```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
...
+        indicesMap[i] = -1;
+    }
+    function getResultItem(dimIdx: number) {
+        const idx = indicesMap[dimIdx];
+        if (idx < 0) {
+            const dimDefItemRaw = dimsDef[dimIdx];
+            const dimDefItem = isObject(dimDefItemRaw) ? dimDefItemRaw : { name: dimDefItemRaw };
+            const resultItem = new DataDimensionInfo();
+            const userDimName = dimDefItem.name;
+            if (dataDimNameMap.get(userDimName) != null) {
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:138]: > Do we need to distinguish user defined string 'null' or 'undefined' from null / undefined here?
I don't understand what you mean

File: src/data/helper/createDimensions.ts [hunk 9]
```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
...
+    }
+    else {
+        // Sort dimensions
+        const toSort = [];
+        for (let i = 0; i < indicesMap.length; i++) {
+            if (indicesMap[i] >= 0) {
+                toSort.push({ i, o: result[indicesMap[i]]});
+            }
+        }
+        toSort.sort((a, b) => a.i - b.i);
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:344]: Only in `SeriesData` will omit the unused dimensions. The `cloneAllDimensionInfo`  used in data transform has no relationship with it. It uses the `dimension define from the source.

File: src/data/SeriesData.ts (hunk 7 above)
Comment (pissang) [reply src/data/SeriesData.ts:912]: Good point. It will be much simpler if we can reuse the same Data instance.

But for `DataStorage`. I'm thinking it should be the opposite. That all methods should be immutable. So if in one series do filter / downSample. It won't affect other Series which share the same DataStorage.

File: src/data/helper/sourceManager.ts [hunk 10]
```diff
@@ -346,8 +350,78 @@ export class SourceManager {
...
+            const upSourceMgr = this._getUpstreamSourceManagers()[0];
+
+            if (isSeries(this._sourceHost) && upSourceMgr) {
+                cachedStore = upSourceMgr._innerGetDataStorage(seriesSource, sourceReadKey);
+            }
+            else {
+                // Always create datastorage based on source from series.
+                const dimensionsDefine = seriesSource.dimensionsDefine;
+                // Can't create a store if don't know dimension..
+                if (seriesSource && dimensionsDefine) {
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/sourceManager.ts:411]: In this case. Dataset will still generate `dimensionsDefine` for the source. `dimensionsDefine` is only undefined when the source is `series.data` instead of dataset. In which case sharing storage is meaningless.

File: src/data/helper/sourceManager.ts [hunk 11]
```diff
@@ -346,8 +350,78 @@ export class SourceManager {
...
+        }
+        const source = this.getSource(0);
+        const dimensionsDefine = source.dimensionsDefine;
+
+        // Source from endpoint(usually series) will be read differently
+        // when seriesLayoutBy or startIndex(which is affected by sourceHeader) are different.
+        // So we use this three props as key.
+        const sourceReadKey = source.seriesLayoutBy
+            + '$$'
+            + source.startIndex
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/sourceManager.ts:382]: fixed

File: src/data/DataStorage.ts [hunk 12]
```diff
@@ -0,0 +1,1281 @@
...
+
+    /**
+     * If we using dataset.
+     * Dimensions types may only know when we initializing series.
+     * So we need to sync the type back to storage when initlializing SeriesData back
+     *
+     * Will return false if dimension type has been known and different from given.
+     * We need to recreate a new data storage in this case.
+     */
+    // TODO Can't sure what's frequency will this validate fail and cause datastorage recreate.
```
Comment (100pah) [inline src/data/DataStorage.ts:308]: ### Strategy thinking

If a category dimension is used in more than one axis, for example two or more grid layout vertically where there is multiple xAxis but they are totally sync and using the same dimension, the entire dataset will not be shared.

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/data/SeriesData.ts (hunk 7 above)
Comment (100pah) [reply src/data/SeriesData.ts:930]: Yes I agree, `DataStorage` will be totally immutable. (The only exception is `appendData` and `appendValue`, which are restricted to only be called in `Series#getRawData`.

File: src/data/helper/sourceManager.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (100pah) [reply src/data/helper/sourceManager.ts]: Use this option:
```js
option = {
    dataset: {
        source: [
            [12, 55, 33],
            [11, 85, 77]
        ]
    },
    xAxis: {},
    yAxis: {},
    series: [{
        type: 'scatter',
        encode: {x: 0, y: 1},
    }, {
        type: 'scatter',
        encode: {x: 0, y: 2},
    }]
};
```
`dimensionsDefine` is `undefined` here. But I've not dig into what's wrong with it yet.

File: src/data/helper/createDimensions.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (100pah) [reply src/data/helper/createDimensions.ts:138]: > We should support user-defined 'null' / 'undefined' to be parsed to null / undefined?

1. I mean add `userDimName != null` before `dataDimNameMap.get(userDimName)`.

2. I mean might be 
```ts
            if (userDimName != null && dataDimNameMap.get(userDimName) == null) { // change != to ==
                resultItem.name = resultItem.displayName = userDimName;
            }

```

File: src/data/helper/createDimensions.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (100pah) [reply src/data/helper/createDimensions.ts:344]: > Only in SeriesData will omit the unused dimensions. The cloneAllDimensionInfo used in data transform has no relationship with it. It uses the `dimension define from the source. 

Oh yes. Do not need to care `cloneAllDimensionInfo`.

Only the test case above should be fixed.

**Comment by pissang** [comment]
@100pah Thanks for the very detailed review. I rethink some of the designs again from your review and did some changes.

Not prepare data storage ahead of `createDimensions`. Instead, we first prepare dimensions in `createSeriesData`(renamed from `createListFromArray`), then pass the dimension to `SourceManager#getSharedDataStorage`(renamed from `SourceManager#getDataStorage`). It will generate a hash from the `type` and `ordinalMeta` in the dimensions. And using this hash to get and cache the DataStorage. 

In this way, `DataStorage` can have enough dimension information for creation. And `SeriesData` can avoid validation and recreation of `DataStorage`.

If we don't use `createSeriesData`(for example in pie it uses createSeriesDataSimply). The strategy is the same as the master branch. We use `getSource` to get the `Source` instance(now it won't return DataStorage, which is a bit weird) and `SeriesData` will be responsible to initialize the DataStorage from this Source. 

So basically this change makes DataStorage sharing not as aggressive as previous. We deal some of the sharing logic in the helper function in `createSeriesData` and use the name `SourceManager#getSharedDataStorage` to make it more clear. It can be expanded to `createSeriesDataSimply` if necessary.

File: src/data/DataStorage.ts (hunk 12 above)
Comment (pissang) [reply src/data/DataStorage.ts]: I've commented about some changes. Now ordinalMeta should be in the key for query and caching OrdinalMeta. No more validation and recreation should happen when ordinaMeta is not the same.(Still there is a `canUse` check to avoid some unexpected scenarios but it should not exist in our cases.

File: src/data/DataStorage.ts [hunk 13]
```diff
@@ -0,0 +1,1281 @@
...
+            if (
+                !selfDim
+                // Dim type can be convert between because ctors are compatitable.
+                || dataCtorCategory[selfDim.type || 'float'] !== dataCtorCategory[targetDim.type || 'float']
+                // ordinalMeta is different. Usually being on the different axis.
+                || (selfDim.ordinalMeta && selfDim.ordinalMeta !== targetDim.ordinalMeta)
+            ) {
+                return false;
+            }
+            selfDim.type = targetDim.type;
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/DataStorage.ts]: It has been removed. Now this method has been changed to `canUse`, only validation check will be prefromed.

File: src/data/DataStorage.ts [hunk 14]
```diff
@@ -0,0 +1,1281 @@
...
+    else {
+        storage[dimIdx] = new DataCtor(end);
+    }
+};
+class DataStorage {
+    private _chunks: DataValueChunk[] = [];
+
+    private _provider: DataProvider;
+
+    // Raw extent will not be cloned, but only transfered.
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/DataStorage.ts]: Removed

File: src/chart/helper/createSeriesDataFromArray.ts [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+    let dimInfoList = createDimensions(sourceOrStore, createDimensionOptions);
+    let firstCategoryDimIndex = injectOrdinalMeta(dimInfoList, opt.createInvertedIndices, coordSysInfo);
+
+    if (omitUnusedDimensions) {
+        // sourceOrStore
+        if (!(sourceOrStore as DataStorage).syncDimensionTypes(dimInfoList)) {
+            // We need a full dimensions list if we want to recreate the storage.
+            // Or the storage will ignore the data of other dimensions.
+            dimInfoList = createDimensions(sourceOrStore, zrUtil.extend(createDimensionOptions, {
+                omitUnusedDimensions: true
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/chart/helper/createSeriesDataFromArray.ts]: Fixed. 

In fact, after fixing this bug, I noticed stacked high dimension data still has a serious performance issue because the stack dimensions will cause `canUse`(syncDimensionsInfo previously) to fail and DataStorage creation. This makes me have to rethink the design.

File: src/data/helper/sourceManager.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/sourceManager.ts]: My mistake, seems dimensionDefine will be undefined if `dimensions` is not given.

File: src/data/helper/createDimensions.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:138]: I split the original logic in https://github.com/apache/echarts/blob/master/src/data/helper/completeDimensions.ts#L115 into two: 

https://github.com/apache/echarts/pull/15355/files/3de73abe262912b928e23b6eab6ac568a67bd7d7#diff-db6362f78d46bd7ee44e35f6bc622f7f7aa6d0a25ec5808f3bf481fbe7a80bb0R160
and
https://github.com/apache/echarts/pull/15355/files/3de73abe262912b928e23b6eab6ac568a67bd7d7#diff-db6362f78d46bd7ee44e35f6bc622f7f7aa6d0a25ec5808f3bf481fbe7a80bb0R139

Because I need to create and set the result dimension only when it's used.

File: src/data/helper/createDimensions.ts [hunk 16]
```diff
@@ -44,33 +51,350 @@ export type CoordDimensionDefinitionLoose = CoordDimensionDefinition['name'] | C
...
+                // Series specified generateCoord is using out.
+                if (!generateCoord || generateCoordCount <= 0) {
+                    resultItem.isExtraCoord = true;
+                }
+                generateCoordCount--;
+            }
+
+            if (resultItem.name == null) {
+                const res = genName(
+                    resultItem.coordDim, dataDimNameMap, dataDimNameAutoIdx, false
```
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:303]: Reversed this change. `genName` should not be a performance issue if we omit the unused dimensions

File: src/data/helper/createDimensions.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:138]: Add a general name duplication remove pass in https://github.com/apache/echarts/blob/690647c7480964e600c0fecfde52738e86f18c93/src/data/helper/createDimensions.ts#L348

File: src/data/helper/sourceManager.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/sourceManager.ts]: I add default dimensions generation in the dataset. It seems more reasonable than picking the dimensions from the coordinate system from the dataset side. But for better display in the tooltip. it will still use `displayName` from coordinate system. Which is 'x', 'y', 'value', etc.

File: src/data/helper/createDimensions.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:343]: Not sure why but seems this test case works as 5.1.2

[the remark this answers is not in the document]
Comment (100pah) [reply src/data/helper/createDimensions.ts:341]: With the same option above, 

In 5.1.2: 
Screen Shot 2021-08-03 at 4 14 20 PM

But in the newest commit 
ae9226ecc01eb60e5662f34c8bcf275dbece17d8 [ae9226ecc]
July 27, 2021 at 5:16:25 PM GMT+8

Screen Shot 2021-08-03 at 4 16 21 PM

[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:341]: @100pah Understand. The rule of picking the last dimension as value is broken.

[the remark this answers is not in the document]
Comment (pissang) [reply src/data/helper/createDimensions.ts:341]: @100pah Fixed it. The original `SeriesData#getDimensionInfo().index` is replaced by `SeriesData#getDimensionIndex`.
Which is always correct for reading `SeriesData#getRawValue`.

Currently the relationship between `SeriesData#dimensions` and `SeriesData#getStorageDimensions` is a bit confusing and I commented to explain it. But still not sure if we will use it wrong further.

[the remark this answers is not in the document]
Comment (100pah) [reply src/data/helper/createDimensions.ts:341]: The index of `SeriesData['dimensions']` and `SeriesData['getDimension']` is not the dimensionIndex that users known, but the internal dimension index if dimension count > 30. 
But there are someplace still output that index to users:
+ in `CustomView`: `wrapEncodeDef(customSeries.getData())`), 
or use user input dimension index on `SeriesData['dimensions']`/`SeriesData['getDimension']`:
+ d
And `RadarView` has bug if use `encode` before this PR.
```js
```


    const userParams: CustomSeriesRenderItemParams = {
        // The life cycle of context: current round of rendering.
        // The global life cycle is probably not necessary, because
        // user can store global status by themselves.
        context: {},
        seriesId: customSeries.id,
        seriesName: customSeries.name,
        seriesIndex: customSeries.seriesIndex,
        coordSys: prepareResult.coordSys,
        dataInsideLength: data.count(),
        encode: wrapEncodeDef(customSeries.getData())
    } as CustomSeriesRenderItemParams;

**Comment by 100pah** [comment]
Some changes has been made in dffeb3a9a


## There are these issues existing before this commit:
1. If no dimensions specified on dataset, series can not really share one storage instance. (Because the each series will create its own dimension name (like ['x', 'y', 'value'], ['x', 'value', 'y'], and the storage hash is based on those names. So the hash can not match).
2. Each time `setOption` update series (but not change `dataset`), new data stack dimensions (and corresponding chunks) will be keep added to the shared data storage, and then the chunks will be more and more.
3. When "unused dimension omit" happen, the index of SeriesData['dimensions'] and SeriesData['getDimension'] is not the dimensionIndex that users known. But there are someplace still use dimensionIndex to visit them. (especially in visualMap and other similar cases that user can input dimension index via option).
4. If user only specify type but no name in dimensions, their will be some bug when "unused dimension omit" happen. (Because unused dimensions will not auto-generate dimension name by `createDimensions` and so that it has no dimension name in storage, and can not be queried by dimension name).
5. If different series option specify its own `dimensions` but share one `dataset`, the `source` get by `sourceManager` is different `source` instances in each series. Those `source` instances contain different `dimensionDefine` but reference the same data. And then a data storage created by based on s

## This commit try resolve those issues by this way:
1. Do not save the "dimName->dimIndex map" in data storage any more. Because:
    1. In fact data storage do not need this map to read/write data.
    2. dimNames are usually created based on each series info (like ['x', 'y', 'value'], ['x', 'value', 'y'], ...) if not specified by user. So they are different between series. But even those series have different generated dimension names, they can still share one storage, because essentially they visit the same dataset source by dimIndex.
2. Make `SeriesDimensionDefine` (that is, each item of `SeriesData['dimensionInfos']`) contain `storageDimensionIndex` to indicate its corresponding data store dimension index. And alway user `storageDimensionIndex` to visit dat storage rather than dimName. `storageDimensionIndex` is created in `createDimension`.
3. create a new structure `SeriesDimensionRequest` for each series. It contains the info generated by `createDimension` (like dimensionDefineList, whether dimension omitted, source for this series).
    3.1. `sourceManager` use `seriesDimensionRequest` to find the shared storage by generate storage dimensions and hash based on the `dimCount` and `dimensionDefineList` (which are created by `createDimension` and saved in `seriesDimensionRequest`).
    3.2. `dataStack` add "data stack dimensions" to `dimensionDefineList` in `seriesDimensionRequest`.
    3.3. `seriesData` use `seriesDimensionRequest` to init its dimensions, and use `seriesDimensionRequest` query dimName by dimIndex from source, or query dimIndex by dimName from source for "omitted dimension". If different series option specify its own `dimensions` but share one `dataset`, the `source` get by `sourceManager` is different `source` instances in each series. Those `source` instances contain different `dimensionDefine` but reference the same "raw data". The data storage generated based on the "raw data" can be shared between series, but the `dimensionDefine` should not be shared. `SeriesDimensionRequest` encapsulate these mess and work each series to query dimension name or index.
    3.4. `seriesDimensionRequest` do not create new data structure as possible as it can, but reference shared data structure (like `source` instance). So it will not cost memory issue.
4. Change the previous `storage.appendDimension` to `storage.ensureCalculationDimension` for data stack. That is, if its dimension has been created, reuse it.
5. Remove previous `canUse` method. Whether a storage can be shared by series is all determined by hash. The hash is generated in two ways:
    5.1. For source format "arrayRows" (i.e., [[12, 33], [55, 99], ...]), dimension name do not need to be added to hash, because this kind of data actually visited by index. If two series have different dimension name (like 'x', 'y') for single index, they also can share the storage.
    5.2. For source format "objectRows" (i.e, [{a: 12, b: 33}, {b: 55, a: 99}, ...]), property name 'a', 'b' will be added to hash, because this kind of data actually visited by property name.
    5.3. And as before, dimension type, ordinal meta id, source header, series layout will also be added to hash.
6. Make `DataStorage` method immutable:
    `DataStorage['filterSelf']` -> `DataStorage['filter']`
    `DataStorage['selectRange']`

## PENDING:
1. Should deprecate `dimName = seriesData.getDimension(dimLoose)` and `series.get(dimName)` but always use `dimIdx = seriesData.getDimensionIndex(dimLoose)` and `dataStorage.get(dimIdx)` instead. For examples:
```js
// Previously
const val = data.get(data.getDimension(dim), dataIdx);
// Now
const val = data.getStorage().get(data.getDimensionIndex(dim), dataIdx);
```
`seriesData.getDimension(dimLoose)` has a feature that convert dimIdx to dimName, which is not essentially necessary (because dimIdx can be used to visit data directly), but this feature require a "dimIdx->dimName map" in `SeriesData` (why? because when some dimensions are omitted, we can not use dimIdx on `SeriesData['dimensions']` directly).

2. Radar has bug when using `series.encode`. This commit do not fix the bug but keep as it is.

**Review by 100pah — APPROVED** [review]
